### ISSUE TITLE
Additional column for investigation and dataset sizes

### ISSUE THREAD

**Issue by gc-code** [icatproject/icat.server#211]
Currently, when a request is made to get the size of an investigation or dataset, a query grabs all the data files associated with the entity and sums their sizes. Seeing as the size of each file is computed at the time of ingestion, the capability should be provided to do the same on the investigation and dataset level. Otherwise effort is being unnecessarily duplicated in performing JOIN operations. This would provide a significant time saving to the end user.

This would also reduce the need for Topcat to maintain its own cache, resulting in fewer inconsistencies between the Topcat cache and the actual data in ICAT.

**Comment by brianritchie1312** [comment]
Topcat caches getSize values both in its server side and in the browser. I think that the browser cache exists to reduce what would otherwise be a flood of requests between the (possibly many) browsers and the topcat server; if so, then it is still worth doing even if the ICAT/IDS response to the request can be sped up. Similar reasoning could be applied to topcat's server-side cache: it reduces the volume of requests from topcat to ICAT. So although this change would probably improve the performance of individual getSize requests, I don't think it means we can remove caching in Topcat.

**Comment by RKrahl** [comment]
I'm mot sure if this brings more benefit then degradation. It may improve the speed of querying the sizes, but it would on the other hand significantly raise the cost of ingestion.

Note that you don't need (and should not) query the individual files related to an investigation and do the summation on the client side. You should rather do a query like `SELECT SUM(df.fileSize) from Datafile df WHERE df.dataset.investigation.id = ?`. This pushes all the calculation into the database backend, which should perform reasonably. This is what the `getSize` ids.server call does internally and TopCAT in turn uses this call since version 2.4.1. Does anybody has measured the performance of this on a real big ICAT since?

**Comment by gc-code** [comment]
On Topcat 2.4.1 I measured the time to load the sizes of a set of search results to be 1 minute and 4 seconds. Seeing as sizes are currently computed from the bottom of the table to the top, the user is waiting that long to see the size of their top search result (I am working on a solution to this). I don't think a user will put up with this. An example browse page that isn't cached takes about 25 seconds.

I would like more computation to go onto the ingestion side because that way it is all done offline, rather than pushing the computation to when the user wants to see their data. In any case, the old method could be kept for when there is no stored value in the table.

The current method of speeding getSize calls is for Topcat to keep a large cache in a database table, that I have found is very prone to inconsistencies with the current data in ICAT. The ingestion process will be better informed on the information going into ICAT and is more easily modified by individual facilities.

**Comment by gc-code** [comment]
I would rely on every ingestion client to keep the sizes up to date, if they choose to do so. I don't think this is too much to ask seeing as they are already being relied upon to provide the sizes of individual files.

**Comment by RKrahl** [comment]
I'd suggest that this would be much more likely to become inconsistent then the TopCAT cache.

**Comment by gc-code** [comment]
Thinking about this further, you are right if user uploaded data is allowed. I will look into creating a process that can run periodically and verify the correctness of the Topcat cache, that may be the best way to approach it.

**Comment by brianritchie1312** [comment]
In Topcat, reloading the browser page (at least in Firefox) clears the browser-side cache, so that it must request entity sizes from the topcat server; but the topcat server's cache persists (even across deployments, I believe!) So the very first time a particular set of investigation sizes are requested, Topcat will query the IDS client for each of them; if the page is then reloaded, the getSize requests will be drawn from the topcat server cache. So the difference in response times for these two cases should give an indication of the additional cost of querying the IDS client.
To get an idea of the overhead of the topcat browser/server communication: browse to a page of investigations, then reload the page to clear the browser cache; note how long it takes for the page to load (and all the sizes to appear). Then drill down on a dataset; then use the browser's back button to return to the investigations page. For this latter operation, the investigation sizes should all be in the browser cache, and (I suspect) the page will load more quickly. Of course, the browser itself may be doing some caching of page content too, for all I know; but I tried this when I had added extensive logging to the cache operations, and this showed that when I used the Back button, topcat drew the investigation sizes from the browser-side cache (but queried its server when the page was reloaded).

**Comment by RKrahl** [comment]
I wonder if these additional size columns could be kept up-to-date using SQL triggers.  E.g. creating a trigger in the database backend on the Datafile table that would recalculate and update the size for the related Dataset and Investigation.

Obviously there would be some issues with that idea:

+ we would need to add these triggers directly in the database backend. That means we would need a separate implementation for each of the supported backends and we would need to hope that all these implementations would show some sort of a consistent common behavior.

+ it would need to be compatible with the Java Persistence API that ICAT uses. We would need to verify that JPA does not choke on an object's attributes that magically change their values behind the scenes.

+ One question would be, how to avoid overwriting the calculated value when there is an update call on the Dataset or Investigation object from the client. Maybe one could make this attribute read-only from the client's perspective.

+ we would need to keep an eye on performance.

In short, we would need to experiment on the idea. It's not clear whether it would work. But it might be worth trying out.
